### PR TITLE
vec: add try_* methods and a try_vec! macro to make Vec usable in without infallible allocation methods

### PR DESCRIPTION
**Pull request by dpaoliello**
As a part of the work for #86942 the `no_global_oom_handling` cfg was added in #84266, however enabling that cfg makes it difficult to use `Vec`: the `vec!` macro is missing and there is no way to insert or push new elements, nor to create a `Vec` from an iterator (without resorting to `unsafe` APIs to manually expand and write to the underlying buffer).

This change adds `try_` equivalent methods for all methods in `Vec` that are disabled by `no_global_oom_handling` as well as a `try_vec!` as the equivalent of `vec!`.

Performance and implementation notes: A goal of this change was to make sure to NOT regress the performance of the existing infallible methods - a naive approach would be to move the actual implementation into the fallible method, then call it from the infallible method and `unwrap()`, but this would add extra compare+branch instructions into the infallible methods. It would also be possible to simply duplicate the code for the fallible version - I did opt for this in a few cases, but where the code was larger and more complex this would lead to a maintenance nightmare. Instead, I moved the implementation into an `*_impl` method that was then specialized based on a new `VecError` trait and returned `Result<_, VecError>`, this trait also provided a pair of methods for raising errors. Never (`!`) was used as the infallible version of this trait (and always panics) and `TryReserveError` as the fallible version (and returns the correct error object). All these `VecError` method were marked with `#[inline]`, so after inlining the compiler could see for the infallible version always returns `Ok()` on success (and panics on error) thus the `?` operator or `unwrap()` call was a no-op, and so the same non-branching instructions as before are generated.

I also added `try_from_iter` and `try_expand` methods for completeness, even though their infallible equivalents are trait implementations.

List of Vec APIs added:
```rust
try_vec!

impl<T> Vec<T> {
    pub fn try_with_capacity(capacity: usize) -> Result<Self, TryReserveError>;
    pub fn try_from_iter<I: IntoIterator<Item = T>>(iter: I) -> Result<Vec<T>, TryReserveError>;
}

impl<T, A: Allocator> Vec<T, A> {
    pub fn try_append(&mut self, other: &mut Self) -> Result<(), TryReserveError>;
    pub fn try_extend<I: IntoIterator<Item = T>>(&mut self, iter: I, ) -> Result<(), TryReserveError>;
    pub fn try_extend_from_slice(&mut self, other: &[T]) -> Result<(), TryReserveError>;
    pub fn try_extend_from_within<R>(&mut self, src: R) -> Result<(), TryReserveError> where R: RangeBounds<usize>; // NOTE: still panics if given an invalid range
    pub fn try_insert(&mut self, index: usize, element: T) -> Result<(), TryReserveError>; // NOTE: still panics if given an invalid index
    pub fn try_into_boxed_slice(self) -> Result<Box<[T], A>, TryReserveError>;
    pub fn try_push(&mut self, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize(&mut self, new_len: usize, value: T) -> Result<(), TryReserveError>;
    pub fn try_resize_with<F>(&mut self, new_len: usize, f: F) -> Result<(), TryReserveError> where F: FnMut() -> T;
    pub fn try_shrink_to(&mut self, min_capacity: usize) -> Result<(), TryReserveError>;
    pub fn try_shrink_to_fit(&mut self) -> Result<(), TryReserveError>;
    pub fn try_split_off(&mut self, at: usize) -> Result<Self, TryReserveError> where A: Clone; // NOTE: still panics if given an invalid index
    pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
}

#[doc(hidden)]
pub fn try_from_elem<T: Clone>(elem: T, n: usize) -> Result<Vec<T>, TryReserveError>;

#[doc(hidden)]
pub fn try_from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Result<Vec<T, A>, TryReserveError>;

```

Other helper APIs added, than can be made `pub(crate)` if folks prefer:
```rust
impl TryReserveError {
    pub fn alloc_error(layout: Layout) -> Self;
}

impl<T> Result<T, !> {
    pub fn unwrap_infallible(self) -> T;
}
```

### CODE REVIEW

File: library/alloc/src/macros.rs [hunk 1]
```diff
@@ -51,6 +51,70 @@ macro_rules! vec {
...
+    () => (
+        $crate::__rust_force_expr!(core::result::Result::Ok($crate::vec::Vec::new()))
+    );
+    ($elem:expr; $n:expr) => (
+        $crate::__rust_force_expr!($crate::vec::try_from_elem($elem, $n))
+    );
+    ($($x:expr),+ $(,)?) => (
+        $crate::__rust_force_expr!({
+            let values = [$($x),+];
+            let layout = $crate::alloc::Layout::for_value(&values);
```
Comment (dpaoliello) [inline library/alloc/src/macros.rs:110]: I really don't like that creating this `layout` is necessary, however the `TryReserveError` type requires a layout for allocation errors, but the `AllocError` type returned by `Box::try_new` does not have this information.

Either there should be some way to get the layout from `AllocError` (which would probably be a breaking change), or `TryReserveError` should not have a `layout` field.

File: library/alloc/src/vec/mod.rs [hunk 2]
```diff
@@ -609,6 +749,51 @@ impl<T, A: Allocator> Vec<T, A> {
...
+    /// assert_eq!(vec.capacity(), 10);
+    ///
+    /// // ...but this may make the vector reallocate
+    /// vec.try_push(11)?;
+    /// assert_eq!(vec.len(), 11);
+    /// assert!(vec.capacity() >= 11);
+    /// # Ok::<(), alloc::collections::TryReserveError>(())
+    /// ```
+    #[inline]
+    #[unstable(feature = "more_fallible_allocation_methods", issue = "86942")]
```
Comment (dpaoliello) [inline library/alloc/src/vec/mod.rs:792]: What's the policy for cases where an API depends on two unstable features (e.g., in this case `more_fallible_allocation_methods` and `allocator_api`)?

**Comment by dpaoliello** [comment]
FYI @Ericson2314 @Xuanwo

**Comment by Ericson2314** [comment]
Great job trying to tackle the dedup-without-penality problem!

An additional technical I would like to see tried is returning `Result<Self, !>` in the infallible case. Right now, life is easy because we are just doing flat containers. But I originally also did infallible linked lists and other more complicated things, where they are multiple possible failure points.

I think the `Result<Self, !>` will scale better, allowing dedup in that case too. The idea would be to have `Result<Self, Operations::Error>`, not the associated error type. And the most code just does `try!` everwhere, but in the `Error = !` case, that `try` becomes a no-op.

This might seem overkill on this PR, but if folks are willing to give it a try I think it's better to do so sooner rather than later.

**Comment by dpaoliello** [comment]
> Great job trying to tackle the dedup-without-penality problem!
> 
> An additional technical I would like to see tried is returning `Result<Self, !>` in the infallible case. Right now, life is easy because we are just doing flat containers. But I originally also did infallible linked lists and other more complicated things, where they are multiple possible failure points.
> 
> I think the `Result<Self, !>` will scale better, allowing dedup in that case too. The idea would be to have `Result<Self, Operations::Error>`, not the associated error type. And the most code just does `try!` everwhere, but in the `Error = !` case, that `try` becomes a no-op.
> 
> This might seem overkill on this PR, but if folks are willing to give it a try I think it's better to do so sooner rather than later.

I was worried that using `!` as the error type would run into some of the issues noted in #35121 (I don't fully understand what exactly those issues are, or how to detect them).

I've tried switching the code to use `!` locally and it seems to be working - if you think it is a good idea, then I'm happy to push it to this change.

**Comment by Ericson2314** [comment]
@dpaoliello I think the issues in that thread are mainly worrying about arbitrary code. This is a mere implementation detail in `alloc`, and so we don't need to be so nervous.

Maybe keep a branch for the old version in case someone disagrees with me, but please do push! I would be very curious to see what you've got :).

**Comment by dpaoliello** [comment]
> @dpaoliello I think the issues in that thread are mainly worrying about arbitrary code. This is a mere implementation detail in `alloc`, and so we don't need to be so nervous.
> 
> Maybe keep a branch for the old version in case someone disagrees with me, but please do push! I would be very curious to see what you've got :).

Ok, pushed! It's a trivial change, unless I've misunderstood what you were suggesting...

File: library/alloc/src/macros.rs (hunk 1 above)
Comment (Ericson2314) [reply library/alloc/src/macros.rs:110]: It is a good design, because in the general case there multiple allocations that could have gone wrong, and so we want to know which one that did.

File: library/alloc/src/vec/mod.rs [hunk 3]
```diff
@@ -95,59 +94,149 @@ mod drain;
...
+    fn reserve(vec: &mut Vec<T, A>, additional: usize) -> Result<(), Self::Error>;
+    fn reserve_for_push(vec: &mut Vec<T, A>, len: usize) -> Result<(), Self::Error>;
+    fn with_capacity_in(capacity: usize, alloc: A) -> Result<Vec<T, A>, Self::Error>;
+    fn capacity_overflow<TReturn>() -> Result<TReturn, Self::Error>;
+    fn extend_with<E: ExtendWith<T>>(
+        vec: &mut Vec<T, A>,
+        n: usize,
+        value: E,
+    ) -> Result<(), Self::Error>;
+    fn with_capacity_zeroed_in(capacity: usize, alloc: A) -> Result<RawVec<T, A>, Self::Error>;
```
Comment (Ericson2314) [inline library/alloc/src/vec/mod.rs:147]:
```suggestion
    type Error;
    fn alloc_error(layout: Layout) -> Self::Error;
```
I have a feeling this is all you need! When creates the error, the other calls `handle_alloc_error`.

Then you can make private methods on the containers (rather than this trait) than take the `Operations` policy parameter, and do the right thing. Those private methods only call other such private methods, to ensure the policy is consistent. And then public methods wrap to hide the policy variable.

I think wit this strategy you can deduplicate *far* more code.

Comment (dpaoliello) [reply library/alloc/src/vec/mod.rs:147]: Interesting... I'll try this now and let you know how it goes...

Comment (dpaoliello) [reply library/alloc/src/vec/mod.rs]: That definitely simplifies things - it looks much better now!

File: library/alloc/src/vec/mod.rs (hunk 2 above)
Comment (Ericson2314) [reply library/alloc/src/vec/mod.rs:718]: Not sure! I've wondered that myself :).

File: library/alloc/src/vec/source_iter_marker.rs [hunk 4]
```diff
@@ -31,6 +34,34 @@ where
...
+                != mem::size_of::<<<I as SourceIter>::Source as AsIntoIter>::Item>()
+            || mem::align_of::<T>()
+                != mem::align_of::<<<I as SourceIter>::Source as AsIntoIter>::Item>()
+        {
+            // fallback to more generic implementations
+            return SpecFromIterNested::try_from_iter(iterator);
+        }
+
+        Ok(Self::from_iter_marker_impl::<I>(iterator))
+    }
```
Comment (Ericson2314) [inline library/alloc/src/vec/source_iter_marker.rs:57]: Can you dedup this too? I don't think this trait is public, which could help.

File: library/alloc/src/vec/mod.rs [hunk 5]
```diff
@@ -95,59 +96,82 @@ mod drain;
...
 use self::spec_extend::SpecExtend;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_extend;
 
+pub(crate) trait Operations {
+    type Error;
+    fn capacity_overflow() -> Self::Error;
+    fn alloc_error(layout: Layout) -> Self::Error;
+}
```
Comment (Ericson2314) [inline library/alloc/src/vec/mod.rs:141]: Last idea that occurred to me:
```suggestion
pub(crate) trait VecError {
    fn capacity_overflow() -> Self;
    fn alloc_error(layout: Layout) -> Self;
}
```
`Self` is never actually used before, so just making the class directly about errors should make type inference work better.

File: library/alloc/src/vec/source_iter_marker.rs (hunk 4 above)
Comment (dpaoliello) [reply library/alloc/src/vec/source_iter_marker.rs:57]: I've gone back and deduped all of these: it clean up a ton code.

File: library/alloc/src/vec/mod.rs (hunk 5 above)
Comment (dpaoliello) [reply library/alloc/src/vec/mod.rs:141]: Yep, that also cleaned up a lot of code :)

**Comment by Ericson2314** [comment]
This is looking really nice! Really wonderful to see more red appear over time to balance out that green :).

**Comment by dpaoliello** [comment]
@Ericson2314 Can you approve this PR? Or who should we add to have a look at it?

**Comment by Ericson2314** [comment]
I cannot. CC @joshtriplett can we enqueue this to the libs team somehow?

**Comment by ojeda** [comment]
> the `vec!` macro is missing

It is available, but not really usable (I disabled it in Rust for Linux for that reason). I just sent https://github.com/rust-lang/rust/pull/96089 to do so here.

> and there is no way to insert or push new elements,

One could use `set_len` to do so manually, but yeah, these `try_*` methods are sorely needed!

**Comment by Ericson2314** [comment]
Hmm, I am surprised there is any change whatsoever, as I thought with the type parameters the old code should be basically identical post monomorphization.

I wonder if some of the `unwrap` is not getting optimized away or similar? Perhaps we should have an `unwrap_infallible` or similar. That would also allow avoiding the use of some `::<!>`.

**Comment by Ericson2314** [comment]
Ah, @urgau tells me these are compile time regressions. Well, that's unfortunate and I am not sure what good options we have to reduce them, but at least their existence makes some sense.

**Comment by ojeda** [comment]
From a quick try with a trivial case, I need `#[inline(always)]` on the `impl VecError for !` to get the exact same codegen as a manually duplicated implementation. Now, panics should be rare, so it may actually be better to move that chunk of code out. In fact, shouldn't it be non-`inline`, `cold`?

> Perhaps we should have an `unwrap_infallible` or similar.

There is `unwrap_unchecked`, I guess it could be worth checking whether that makes it easier for the optimizer here.

**Comment by Ericson2314** [comment]
`unwrap_unchecked` is kinda sketchy, but it is possible to write `unwrap_infallible` with reguar safe code. See the quite old https://github.com/reem/rust-void/blob/a6e061227f47ba8798b7e828ed0ac4e25382eb15/src/lib.rs#L72-L101

**Comment by Ericson2314** [comment]
A bit out of scope for this PR, but I think we should have those unstable under the `never_type` feature with `!`.

**Comment by ojeda** [comment]
Yeah, having `unwrap_infallible` around for cases like the one here is definitely better than `unwrap_unchecked`, but I think they would both end up showing as `unreachable` to LLVM, so I thought using `unwrap_unchecked` would be an alternative to test if it made a difference for the compile-times etc.

**Comment by dpaoliello** [comment]
@joshtriplett I've done some experiments with using `inline(always)`, `cold` and adding `unwrap_infallible` for `Result<T, !>`, can you please re-run the perf tests?

**Comment by Ericson2314** [comment]
@dpaoliello Note I was being silly -- this is literally no tag to branch on(!) so `unwrap` must be harmless per wise. Now I still like `unwrap_infallible`, but its sole purpose is better type inference (so no need for `::<!>`), not run-time perf.

**Comment by ojeda** [comment]
> I've done some experiments with using `inline(always)`, `cold`

Thanks for making `capacity_overflow` etc. `cold`. However, why `inline(always)`? The `panic!` setup is quite a bunch of instructions that are most often unused; so why do we want to force inlining? Did you see some bad codegen?

**Comment by Ericson2314** [comment]
Should we try removing the `::<!>` now that we have `unwrap_infallible`?

**Comment by dpaoliello** [comment]
> > I've done some experiments with using `inline(always)`, `cold`
> 
> Thanks for making `capacity_overflow` etc. `cold`. However, why `inline(always)`? The `panic!` setup is quite a bunch of instructions that are most often unused; so why do we want to force inlining? Did you see some bad codegen?

@ojeda I figured that forcing inlining would cause the compiler to see that the code was a panic, and so the `?` operator should be a no-op - it was also a leftover from an older version where I was doing more work in these functions rather than simply returning an error or panicking. Now that the return type of these functions is `!`, it seems that the compiler correctly sees that they never return even without the `inline` attribute, and removing that attribute generates much nicer assembly. Thanks!

**Comment by dpaoliello** [comment]
> Should we try removing the `::<!>` now that we have `unwrap_infallible`?

@Ericson2314 Good idea! I was able to get rid of almost all of the explicit types, except in one case: `vec::try_append`, for some reason the compiler couldn't infer the type in that case.

**Comment by dpaoliello** [comment]
> This is quite a large PR to review, and it's not directly obvious what public API this adds. Can you give a full overview of the api surface?

Updated the PR description with the list of APIs added.

**Comment by ojeda** [comment]
Having the signatures etc. could make it easier for reviewers, e.g.

```rust
impl<T, A: Allocator> Vec<T, A> {
    pub fn with_capacity_in(capacity: usize, alloc: A) -> Self;
    pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
}
```

**Comment by dpaoliello** [comment]
> Having the signatures etc. could make it easier for reviewers, e.g.
> 
> ```rust
> impl<T, A: Allocator> Vec<T, A> {
>     pub fn with_capacity_in(capacity: usize, alloc: A) -> Self;
>     pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError>;
> }
> ```
> 
> (and could be copied into the "Public API" section of #86942)

Done

**Comment by dpaoliello** [comment]
@rustbot label: +perf-regression-triaged
Regressions are in `LLVM_module_codegen_emit_obj`, this suggests that the code generated for `Vec` has not regressed (I've manually confirmed this by comparing assembly), but rather that LLVM takes longer to emit the code possibly due to the increased number of generic functions.
An average regression of 0.4% seems reasonable to enable the use of `Vec` within the Linux Kernel (and other places where panic-on-oom is not acceptable).

**Comment by m-ou-se** [comment]
The API surface that this PR adds is quite large. It increases the number of methods on `Vec` by more than 30%. It's also not obvious to me that this is the only way to expose such an API. Duplicating/splitting all these functions into two variants doesn't seem great. Also, the `try_` prefix often doesn't refer to allocation failures. I've heard ideas about a `Vec::try_push` that only pushes if there's capacity left (without ever trying to reallocate), for example. Also, `try_insert` still panicking on an out-of-bounds index seems confusing.

I don't think I'm comfortable with approving this change without an RFC to first discuss the API and possible alternatives.

File: library/alloc/src/vec/mod.rs (hunk 2 above)
Comment (m-ou-se) [reply library/alloc/src/vec/mod.rs:715]: See https://github.com/rust-lang/rust/issues/94770

**Comment by ojeda** [comment]
Sounds reasonable.

In fact, an RFC would be good to discuss the design and whether it fits the Linux kernel too.

**Comment by dpaoliello** [comment]
> The API surface that this PR adds is quite large. It increases the number of methods on `Vec` by more than 30%. It's also not obvious to me that this is the right way to expose such an API. Duplicating/splitting all these functions into two variants doesn't seem great. Also, the `try_` prefix often doesn't refer to allocation failures. I've heard ideas about a `Vec::try_push` that only pushes if there's capacity left (without ever trying to reallocate), for example. Also, `try_insert` still panicking on an out-of-bounds index seems confusing.
> 
> I don't think I'm comfortable with approving this change without an RFC to first discuss the API and possible alternatives.

I'm not sure that I'm the correct person to write such an RFC (@Ericson2314 would you be willing to write such a doc?): I'm more than happy to do the work to implement proposals, but I don't have enough experience with Rust to know what type of API design is going to feel most ergonomic to Rust users.

It also feels bizarre to me that the existing RFC for fallible allocation (https://github.com/rust-lang/rfcs/pull/3140) suggests adding in a feature/config to exclude infallible functions without describing what the fallible functions look like...

I agree that adding a bunch of `try_` functions is not a great design: it blows up the API surface, and that's before having to bifurcate a bunch of traits so that things like `collect()` are also useful.

The design for fallible vs infallible functions was also discussed in RFC 2116: https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md#future-work-complete-result-apis

I'm not sure how great the alternatives are:
* We could change the existing APIs to return `Result` (which would be a breaking change) and have the error type be an associated type of the allocator (and so `Global` would have an error type of `!`, which is going to be confusing to teach new developers).
  * To avoid the breaking change and new developer confusion, we could move the implementation of `Vec` into a new type that has `Result`-returning methods based on the allocator being used, then change `Vec` itself to be a simple wrapper around that type keeping the current API surface.
* Allocation failures could be out-of-band on the allocator itself (i.e., you'd need to manually check the allocator after every API call), but that seems verbose and error-prone.

**Comment by Ericson2314** [comment]
@m-ou-se yeah when we discussed the first bunch of `try_` methods we agreed no on liked `try_` everwhere, but as @dpaoliello mentions, the alternatives we can do in the short term are worse. The decision then was IIRC to try to do enough `try_` to get Rust in Linux off a functionality-adding fork, but never *stabilize* such a doubling of the interface.

I still think that's a good idea. Long term, I'm optimistic we'll get something to help here, say a varistion on editions or the standard library that won't split the ecosystem, or a magic `Result<T, !> -> T` coercion, that will shake up the stalemate. But, I don't think that holding this new functionality "hostage" until we do is a good idea. I would rather plug our nose, merge this, and then *leverag* the ugliness of the API to prioritize getting one of those long term solutions.

When I was first trying to get the no global OOM handling hack merged, I was saying I was worried the Linux side wouldn't even consider using liballoc a goal, causing a needless ecosystem split between code for Linux and code for elsewhere. Now that we have that, I think the situation is better, and everyone is on board with the idea of making Linux able to use a vanilla liballoc long term.

(To be clear on meta note, I don't want to appear like a dog that was thrown bone (with that no global OOM handling stop-gap cfg option) who just unsatiated for another. What we have is indeed better than before. I just want to continue down that path and make it better still.)

So, if everyone now agrees getting Linux and a vanilla liballoc is the goal, what's the problem with just being on a fork until then, or rather what's to be gained by being on less of a fork if it's just using a perma-unstable feature? Well, here is one example.

A few weeks back (I'll try to find link later when I'm not on my phone), there was some discussion of being able to use Crates.io crates in Linux. I can't imagine the upstream kernel would so such a thing any time soon, but 3rd party kernel modules may be more adventurous. To me, this would be a very positive development. A major goal of Rust is to make systems programming more accessible, and I'm a true believer that merely by removing enough footguns and ambient authority (deeply related concepts) regular Rust devs writing code for userland may well "accidentally" write code that is valid for the kernel too.

So long as Rust is on a fork with additional features, it would be harder to foster a non-Linux-specific ecosystem of libraries usable by adventurous downstream Kernels / Kernel modules. To be fair, any library that is core-only and doesn't need to dynamically allocate at all would be fine, but any that does need to do dynamic memory allocation is liable to get in more trouble if it needs any functionality whose `Result`-returning form is Linux-only.

Using perma-unstable features is no fun, but it at least works just fine with Linux or non Linux nightly rust alike. I also think that when we *do* come up with such a long term solution to replace `try_` everwhere, it will probably be similar enough to the unstable feature that converting from one to the other would be a rote task. This is why I don't think "Perma unstable" can or should scare people off, nor would there be risk of a critical mass of folks demanding `try_` be standardized as-is (always a risk with unstable features that get too popular).

So in conclusion, I think there are many benefits, and the costs are not so bad as they first appear.

**Comment by rinon** [comment]
I would add onto this that the Linux kernel is not the only place people care about this. I work on Rust for Trusty, an open-source embedded OS for the ARM trusted-execution environment, and we need first-class fallible allocation support. We are using upstream crates, although I'm happy to put in the work to get upstream support for fallible allocation into these crates if we can standardize on something.

**Comment by m-ou-se** [comment]
I think this discussion shows why an RFC would be nice. Then we can have this discussion separately from a thread/PR about implementation details and performance.

**Comment by dpaoliello** [comment]
@Ericson2314 @ojeda @m-ou-se Ok, RFC created: https://github.com/rust-lang/rfcs/pull/3271

I tried to think of as many alternatives as possible. I also wrote down the starting of an idea to deal with "function variations" in the Future Possibilities section, but it would require a lot more language design and discussion.

**Comment by dpaoliello** [comment]
@m-ou-se I'd like to be able to get these functions merged in to unblock folks that need them: if you're worried about cluttering `Vec`, then I'm more than happy to mark them as `#[doc(hidden)]` for now.

**Comment by the8472** [comment]
> Also, the `try_` prefix often doesn't refer to allocation failures. I've heard ideas about a `Vec::try_push` that only pushes if there's capacity left (without ever trying to reallocate), for example.

In #89123 I chose a name for that function that doesn't start with `try_` to avoid a name collision since that prefix already is used for fallible allocation in `try_reserve`.

**Comment by i509VCB** [comment]
For the `try_` functions which add some T or an iterator of T to the Vec, do we need a way to get the value(s) back if allocation fails?

Using the OS kernel example, if pushing to an vec fails in kernel, you may need to take a moment to clean up some garbage/kill some processes and then try again.

**Comment by Ericson2314** [comment]
@i509VCB I think the right way to do that is a more "emplace"-type mechanism by which

1. One ensures the capacity is large enough, and on success gets back a thing like a lock guard
2. One uses the thing like a lock guard to push the item without possibility of failure.

We can do this either way, but the benefits are a lot easier to understand with `Result` returning (i.e., this RFC) because the types (`Result needed for step 1 but not step 2) make explicit what is going on.

**Comment by JohnCSimon** [comment]
ping from triage:
@dpaoliello 
Returning to author to address comments + merge conflicts


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

@rustbot author

**Comment by the8472** [comment]
While most of the APIs at least appear somewhat (but not ideal or low-abstraction-orthogonal-building-blocks)

these two seem more dubious than the others:

```
    pub fn try_from_iter<I: IntoIterator<Item = T>>(iter: I) -> Result<Vec<T>, TryReserveError>;
    pub fn try_extend<I: IntoIterator<Item = T>>(&mut self, iter: I, ) -> Result<(), TryReserveError>;
```

They basically implement inherent versions of `FromIterator` and `Extend` impls. They also lead to a lot of trait impls getting modified. But neither of those interfaces are well-suited for fallible operations because they consume an iterator, take any number of items from it and then may suddenly fail and drop those items - and in the former case a partially allocated vec - on the floor. That is fine in the infallible case where the unwind would usually lead to those getting dropped anyway. But in a case where one needs to recover from errors that can be a bad idea.

Fallible iterator-absorbing methods may need separate implementations that allocate ahead of time (at the risk of allocating a wee bit too much) and hand you back the pieces when stuff breaks.

Removing those would also shrink the diff.

**Comment by dpaoliello** [comment]
Closing this for now.
In the short term, I've open sourced the `FallibleVec` trait we've been using at Microsoft: https://crates.io/crates/fallible_vec
In the long term, hopefully Keyword Generics will provide a solution for this: https://blog.rust-lang.org/inside-rust/2022/07/27/keyword-generics.html

File: library/alloc/src/vec/mod.rs [hunk 6]
```diff
@@ -95,59 +96,74 @@ mod drain;
...
-#[cfg(not(no_global_oom_handling))]
 mod spec_from_iter;
 
-#[cfg(not(no_global_oom_handling))]
 use self::spec_extend::SpecExtend;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_extend;
 
+pub(crate) trait VecError {
```
Comment (Ericson2314) [inline library/alloc/src/vec/mod.rs:137]: @workingjubilee this is mechanism that allows the eager bail out (just like today) or error throwing.
